Fetch reward account balance when reading or listing wallets

Anviking opened the pull request:
# Issue Number

#904 


# Overview

- [x] Byron wallets no longer have a `reward`-field in their `balance`.
- [x] Fetch reward balance when reading or listing Shelley wallets:
    - [x] I have added a `HasRewardAccount` an instance for `SeqState`.
    - [x] I added `getAccountBalance` to `NetworkLayer` which calls `getAccountState` of `JormungandrClient`
    - [x] I added `Cardano.Wallet.fetchRewardBalance` calling `getAccountBalance` with some error-handling.
    - [x] I made sure created Api Wallets ask `fetchRewardBalance` for the reward balance.
- [x] I required the controlled-stake of the pool to increase in `STAKE_POOL_JOIN_01`
- [x] Commented integration test checking that the `rewardBalance` increases

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1181,16 +1181,20 @@ natural = Quantity . fromIntegral . getQuantity
-            \wrk -> W.readWallet wrk wid
+            \wrk -> do
+                (a,b,c) <- W.readWallet wrk wid
+                r <- withExceptT (error "TODO") $ W.fetchWalletRewardBalance wrk wid
```
<s>This doesn't compile when I require `HasNetworkLayer t ctx` from `W.fetchWalletRewardBalance `</s> 🤔 

Ok, explicit type application helped.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -468,6 +470,31 @@ readWallet ctx wid = db & \DBLayer{..} -> mapExceptT atomically $ do
+
+-- Rather than force all callers of 'readWallet' to wait for fetching the
+-- account balance (via the 'NetworkLayer'), we expose this function for it.
+fetchWalletRewardBalance
```
**[suggestion]**

Drop `Wallet` from the name

**Why**

There's no much ambiguity about where the reward balance comes from. It makes the overall clearer.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -671,9 +698,9 @@ listAddresses ctx wid = db & \DBLayer{..} -> do
     -- to make sure that we compare them correctly.
     normalize :: k 'AddressK XPub -> Address -> Maybe Address
-    normalize rewardAccount addr = do
+    normalize rewardAcc addr = do
```
**[change required]**

Fix the top-level signature by removing `s ~ SeqState n k` and using `hasRewardAccount s` instead

**Why**

As the `FIXME` mentions, this was an ugly hack that would better be tackled by using a better abstraction.

Anviking replied:
`rewardAccount` returns a `Maybe ChimericAccount`, and we need a `k 'AddressK XPub` here.

We could change `rewardAccount` to be `Maybe (k 'AddressK XPub)`, but we'd need to involve `k` in the `HasRewardAccount s` class… 🤔

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1199,15 +1203,16 @@ mkApiWallet
+    -> (Quantity "lovelace" Word64)
     -> ApiWallet
-mkApiWallet wid wallet meta progress pending = ApiWallet
+mkApiWallet wid wallet meta progress pending rewards = ApiWallet
```
**[question]**

Why a plural form here :grimacing:  ?

Anviking replied:
Without `balance` as suffix, I thought `rewards` sounded better than `reward`. (@jonathanknowles ?)

`rewardBal[ance]` might be a safe bet though.

jonathanknowles replied:
Either "reward" or "rewards" are fine, I think!

However, perhaps "rewards" has the unwanted connotation of implying a list or some other container.

`rewardBalance` is also a good compromise.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs:
```diff
@@ -232,3 +233,6 @@ instance CompareDiscovery (RndState n) where
     knownAddresses s = Map.elems (addresses s)
+
+instance HasRewardAccount (RndState n) where
+    rewardAccount = error "TODO"
```
**[change required]**

Replace `error "TODO"` with `Nothing`

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -468,6 +471,40 @@ readWallet ctx wid = db & \DBLayer{..} -> mapExceptT atomically $ do
+        r <- liftIO $ runExceptT x
+        case r of
+            Right v -> return v
+            Left _ -> return $ Quantity 0
```
**[change required]**

Do not silently swallow an error. If one error deserves to be ignore, use a clear pattern match to show it but avoid holes. 

**Why**

There are two reasons, the first one is:

- https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#prefer-pattern-matching-to-equality-testing-on-sum-types

And the second is that, if the account doesn't exist on the blockchain, we may want to default to a value of `0`. However, in case we get any sort of error, we should either fail, or, at the very least, log an error.
Note also that this is also an abuse of the `liftE2` function IMO which does much more than just lifting the error. I'd suggest moving the default handling logic to another function and to compose.

KtorZ commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -682,6 +683,14 @@ balanceTotal =
+    => Lens' s Natural
+rewardBalance = lens
+        (getQuantity . getField @"reward")
+        (\(s,v) -> setField @"reward" (Quantity v) s)
```
**[sugestion]**

You may want to rebase on latest master, this lens now exists as `balanceReward`.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1180,17 +1184,23 @@ natural :: Quantity q Word32 -> Quantity q Natural
+    rewards <-
+        liftHandler $ withWorkerCtx ctx wid (throwE . ErrFetchRewardsNoSuchWallet) $
+            \wrk -> do
+                    W.fetchWalletRewardBalance @_ @s @k @t wrk wid
```
**[suggestion]**

Remove the extra `do` and newlines.

**Why**

It creates needless noise in the review, as well as cluttering a bit the function itself. Unless there's a good reason to do so?

Anviking committed:
Add getAccountBalance to NetworkLayer

Anviking committed:
Add (HasRewardAccount s) class

with implementations for SeqState

Anviking committed:
Add fetchRewardBalance in Cardano.Wallet

Anviking committed:
Add commented integration test for accumulation of rewards

And require that the controlled stake of the pool increases for
the existing test.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs:
```diff
@@ -232,3 +233,6 @@ instance CompareDiscovery (RndState n) where
     knownAddresses s = Map.elems (addresses s)
+
+instance HasRewardAccount (RndState n) where
+    rewardAccount _ = Nothing
```
I'm not sure about the `Has`-prefix. Seeing the constraint `HasRewardAccount s` I would like to believe that `s` **has** a reward account. But it doesn't mean that! `rewardAccount @s` could return `Nothing`!

KtorZ replied:
Why could it though ? I don't think `HasRewardAccount` should be defined on `RndState` in the first place actually.

KtorZ replied:
That's kinda the point of all these instances and the way the wallet engine works. You only define the capabilities on the relevant concrete types. 

`RndState` doesn't implement `DelegationAddress` for instance, and that's very fine :)

Anviking replied:
> I don't think HasRewardAccount should be defined on RndState in the first place actually.

No, I agree. (conceptually)

It simplified things in `Api.Server`. I will see now if I can do anything.

Anviking replied:
Changed

Anviking committed:
Add ApiByronWalletBalance type (without reward field)

Anviking committed:
Call fetchRewardBalance in API Server (for Shelley wallets)

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -976,14 +970,14 @@ signDelegation ctx wid argGenChange pwd coinSel poolId action = db & \DBLayer{..
                 putCheckpoint (PrimaryKey wid) (updateState s' cp)
 
-            let rewardAccount = deriveRewardAccount @k pwd xprv
+            let rewardAcc = deriveRewardAccount @k pwd xprv
```
This is a private key. Maybe we should rename `HasRewardAccount.rewardAccount` to `rewardAccountXPub` or something.

KtorZ replied:
`rewardAccountXPub` is redundant with the type signature. Instead, we could make `deriveRewardAccount` part of the `HasRewardInterface` to make sure they're done consistently. Pretty minor here however.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1232,6 +1233,12 @@ getWalletBalance wallet pending = ApiT $ WalletBalance
+getByronWalletBalance :: Wallet s -> Set Tx -> ApiByronWalletBalance
+getByronWalletBalance wallet pending = ApiByronWalletBalance
+    { available = Quantity $ availableBalance pending wallet
+    , total = Quantity $ totalBalance pending wallet
```
Not sure if we have to update swagger.yaml.

KtorZ replied:
It does need to be updated yes: https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/getByronWallet

Anviking replied:
f8c63cc

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -658,6 +660,41 @@ deleteWallet ctx wid = db & \DBLayer{..} -> do
+        Left (ErrGetAccountBalanceAccountNotFound _) ->
+            Right $ Quantity 0
+        Left (ErrGetAccountBalanceNetworkUnreachable e) ->
+            Left $ ErrFetchRewardsNetworkUnreachable e
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -672,11 +709,7 @@ listAddresses
-            -- Perhaps rely on a proper abstraction for normalizing addresses.
-            -- This is _acceptable_ for now as only sequential addresses can be
-            -- listed, but it's not quite elegant nor extensible.
+        , HasRewardAccount s k
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -487,26 +489,34 @@ getWallet
     wids <- liftIO $ Registry.keys re
     fmap fst . sortOn snd <$>
-        mapM (getWalletWithCreationTime mkApiWallet ctx) (ApiT <$> wids)
+        mapM mkWallet wids
```
**[suggestion]**

rewrite to `mapM (getWallet ctx) (ApiT<$> wids)` 

**Why**

Avoid logic duplication in `mkWallet`

Anviking replied:
Unfortunately doesn't work since `listWallets` needs to sort on the creation time in `(ApiWallet, UTCTime)`, while `getWallet` only can return `ApiWallet`.

Maybe we can rewrite some of the `getWalletWithCreationTime` and `mkWallet`-indirection, however…

Anviking replied:
06b04a5

I removed `mkApiWallet` and `mkApiByronWallet` which were arguments to `getWalletWithCreationTime`.

I merged `getWalletWithCreationTime` with `mkApiWallet` and added `getByronWalletWithCreationTime`.

This way there is less indirection.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -267,6 +268,7 @@ spec = do
             jsonRoundtripAndGolden $ Proxy @(ApiTransaction 'Testnet)
             jsonRoundtripAndGolden $ Proxy @ApiWallet
             jsonRoundtripAndGolden $ Proxy @ApiByronWallet
+            jsonRoundtripAndGolden $ Proxy @ApiByronWalletBalance
```
:+1:

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -168,10 +168,36 @@ spec = do
+        eventually $ do
+            request @[ApiStakePool] ctx listStakePoolsEp Default Empty >>= flip verify
+                [ expectListItemFieldSatisfy 0 (metrics . stake)
+                    (> (100000000000 + 1 - 100))
```
**[change required]** 

Use `faucetUtxoAmt` (from Test.Integration.Framework.DSL) and `stakedelegationfee` (from Test.Integration.Framework.TestData)

**Why**

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#prefer-named-constants-over-magic-numbers

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -168,10 +168,36 @@ spec = do
+            request @[ApiStakePool] ctx listStakePoolsEp Default Empty >>= flip verify
+                [ expectListItemFieldSatisfy 0 (metrics . stake)
+                    (> (100000000000 + 1 - 100))
+                ]
```
**[change required]**

Move this expectation into a separate scenario

**Why**

This is yet a different scenarios that would deserve its own test for clarity and debugging. Try to keep scenarios focused on doing one thing.

Anviking replied:
Done. We should perhaps be mindful about that these tests take quite a lot of time though.

Anviking committed:
Update swagger.yaml with new byron wallet balance

Anviking committed:
Use type-family instead of functional dependencies

Anviking committed:
Remove indirection for getWalletWithCreationTime

KtorZ reviewed (commented):
LGTM. I'll look into enabling the remaining integration tests on top of rc4 before merging.

KtorZ committed:
make sure rewards accumulate correctly in the wallet

KtorZ committed:
do not list stake pools when not necessary: use arbitrary id instead

KtorZ committed:
remove superfluous debug lines
